File: generated_code/cpp/solution_1895_2.cpp
#include <vector>
#include <string>
#include <cmath>

using namespace std;

class Solution {
public:
    vector<int> minOperations(string boxes) {
        int n = boxes.length();
        vector<int> answer(n, 0);
        
        // This problem can be optimized to O(n) time using a prefix/suffix accumulation strategy.
        // Instead of calculating the distance from every ball to every box independently (O(n^2)),
        // we can track the accumulated operations dynamically by iterating from left to right, 
        // and then from right to left.
        
        // Pass 1: Left-to-Right
        // Calculate operations needed to bring all balls to the left of index i to index i.
        int count = 0; // Number of balls seen so far from the left
        int ops = 0;   // Accumulation of operations to bring those balls to the current index
        for (int i = 0; i < n; ++i) {
            answer[i] += ops;
            if (boxes[i] == '1') {
                count++;
            }
            ops += count; // Each ball seen so far takes 1 more step to get to the next index
        }
        
        // Pass 2: Right-to-Left
        // Calculate operations needed to bring all balls to the right of index i to index i.
        count = 0; // Reset number of balls seen so far from the right
        ops = 0;   // Reset accumulation of operations
        for (int i = n - 1; i >= 0; --i) {
            answer[i] += ops;
            if (boxes[i] == '1') {
                count++;
            }
            ops += count; // Each ball seen so far takes 1 more step to get to the next index (moving left)
        }
        
        // Time Complexity: O(n) - Two linear passes over the string.
        // Space Complexity: O(1) auxiliary space (excluding the output vector).
        return answer;
    }
};

File: generated_code/python/solution_1895_2.py
from typing import List

class Solution:
    def minOperations(self, boxes: str) -> List[int]:
        """
        Time Complexity: O(n) - Two passes over the string of length n.
        Space Complexity: O(1) - Only allocating space for the output array.
        
        Algorithm: Prefix and Suffix Sums (Linear Scan)
        We can calculate the number of operations required for each box by spliting
        the movement into two independent directions: left-to-right and right-to-left.
        - In the first pass (left-to-right), we track how many balls are to the left 
          of the current box and accumulate their distances.
        - In the second pass (right-to-left), we do the same for balls to the right 
          of the current box, adding their contributions to the answer array.
        """
        n = len(boxes)
        answer = [0] * n
        
        # Pass 1: Left-to-right (accumulate operations from balls to the left)
        left_balls = 0
        left_operations = 0
        for i in range(n):
            answer[i] += left_operations
            if boxes[i] == '1':
                left_balls += 1
            # Moving to the next box means all existing balls on the left take 1 more step
            left_operations += left_balls
            
        # Pass 2: Right-to-left (accumulate operations from balls to the right)
        right_balls = 0
        right_operations = 0
        for i in range(n - 1, -1, -1):
            answer[i] += right_operations
            if boxes[i] == '1':
                right_balls += 1
            # Moving to the next box (leftwards) means all existing balls on the right take 1 more step
            right_operations += right_balls
            
        return answer